common/src/api/external/error.rs

```diff
@@ -255,32 +255,66 @@ impl From<Error> for HttpError {
     }
 }
 
-impl<T: std::fmt::Debug> From<progenitor::progenitor_client::Error<T>>
+pub trait ClientError: std::fmt::Debug {
+    fn message(&self) -> String;
+}
+
+// TODO this From may give us a shortcut in situtations where we want more
+// robust consideration of errors. For example, while some errors from other
+// services may directly result in errors that percolate to the external
+// client, others may required, for example, retries with an alternate service
+// instance or additional interpretation to sanitize the output error.
+//
+// ***********************************************************************
+// * THIS FROM SHOULD BE REMOVED PRIOR TO SHIPPING TO AVOID LEAKING DATA *
+// ***********************************************************************
+impl<T: ClientError> From<progenitor::progenitor_client::Error<T>>
     for crate::api::external::Error
 {
     fn from(e: progenitor::progenitor_client::Error<T>) -> Self {
-        // TODO this should really be something progenitor provides
-        let status = match &e {
-            progenitor::progenitor_client::Error::CommunicationError(e) => {
-                e.status()
+        match e {
+            // This error indicates a problem with the request to the remote
+            // service that did not result in an HTTP response code, but rather
+            // pertained local (i.e. client-side) encoding or network
+            // communication.
+            progenitor::progenitor_client::Error::CommunicationError(ee) => {
+                Error::internal_error(&format!("CommunicationError: {}", ee))
             }
+
+            // This error represents an expected error from the remote service.
             progenitor::progenitor_client::Error::ErrorResponse(rv) => {
-                Some(*rv.status())
-            }
-            progenitor::progenitor_client::Error::InvalidResponsePayload(e) => {
-                e.status()
+                let message = rv.message();
+
+                if rv.status().is_client_error() {
+                    Error::invalid_request(&message)
+                } else {
+                    Error::internal_error(&message)
```

Comment:
Discussed offline: I think a 503 here should still be a 503 out the front door.

I understand if we don't want to bother being too pedantic here, but there are some other cases that probably shouldn't go straight out: 401 and 403, for example.  It might be safer to have an allow-list of codes that we pass through (maybe: 400, 404, 409, 422, 503) and make everything else a 500?

Reply:
With regard to the allow list: 1. I like it 2. it seems challenging to do robustly without some additional facilities from dropshot. For example, consider a 404 not found:

In the `Error` type it's represented like this:

```rust
ObjectNotFound { type_name: ResourceType, lookup_type: LookupType },
```

What we get from the progenitor/dropshot call is (effectively) a `dropshot::HttpError` which has a `message` field that was populated like this:

```rust
                let message = format!(
                    "not found: {} with {} \"{}\"",
                    t, lookup_field, lookup_value
                );
                HttpError::for_client_error(
                    Some(String::from("ObjectNotFound")),
                    http::StatusCode::NOT_FOUND,
                    message,
                )
```

So: in this `impl From<progenitor::progenitor_client::Error<T>> for Error` if we were to see that the status was 404 we would need to parse out the message field into the components for the `ObjectNotFound` variant. That seems both flaky for services within the nexus repo and completely untenable for external services such as crucible which e.g. populates the message field differently:

```
    Err(HttpError::for_not_found(
        None,
        format!("region {:?} snapshot {:?} not found", p.id, p.name),
    ))
```

All of this is to say that I agree that maintaining structured errors across service boundaries seems useful *and* I think we need some more infrastructure to build that robustly.